feat: Add the ability to read the cl_node_urls from env variables

MoeMahhouk opened the pull request:
## 📝 Summary

This PR adds the ability to parse the cl_node_url passed from an environment variable too.
It would help with the automatic deployment in the TDX yocto images.

This addition does not affect the original behaviour of accepting a single value or an array of values. It adds extra a way to pass the values from an env variable where the values are separated by a comma from a single string.

## 💡 Motivation and Context

The motivation behind this change was triggered when preparing the TDX demo with rbuilder and the rbuilder was crashing because we were passing the CL_NODE_URL as environment variables during deployment from the cloud init conf.
This lead us to hardcode it first just for the sake of the demo but I believe it would be better to allow this to be passed and fetched from an environment parameter too similar to the secrets configs.


## ✅ I have completed the following steps:

* [x] Run `make lint`
* [x] Run `make test`
* [x] Added tests (if applicable)

MoeMahhouk committed:
feat: Add the ability to read the cl_node_urls from env variables

MoeMahhouk commented:
I have ran the make tests too and all green besides one of the tests is failing.
But I checked on the develop branch and the same test is failing. The failing test seems not relevant with those changes

ZanCorDX committed:
More comments (#53)

## 📝 Summary

More comments on:
- bin
- backtest
- utils
- primitives

More will come in following PRs.

Commit:
Fix: Use lock properly in filter_order_by_nonces to do it concurrently (#38)

## 📝 Summary

The current use of Mutex on `nonce_cache` in `filter_order_by_nonces` is
not correct. It causes the function to run sequentially because every
other task will wait for prev one ( both read and write ) complete in
order to do its work.
I change to use RwLock instead. Before the change it takes 18m to fetch
a block with 2k4 orders, and now it takes 18s with 400 concurrent
request.

- And lint the file

## 💡 Motivation and Context

<!--- (Optional) Why is this change required? What problem does it
solve? Remove this section if not applicable. -->

---

## ✅ I have completed the following steps:

* [x] Run `make lint`
* [x] Run `make test`
* [ ] Added tests (if applicable)

ferranbt committed:
feat: Add abstraction for the beacon api client (#34)

This PR adds a small abstraction over the beacon client api and
implements the methods required by the rbuilder.

There are a couple of unit tests that require a running beacon node.
They are ignore right now but we will enable them in a followup PR.

<!--- (Optional) Why is this change required? What problem does it
solve? Remove this section if not applicable. -->

---

* [x] Run `make lint`
* [x] Run `make test`
* [x] Added tests (if applicable)

ferranbt committed:
feat: Initialize mempool dumpster path + test (#14)

## 📝 Summary

Builds on top of https://github.com/flashbots/rbuilder/pull/13. This PR
moves the initialisation of the mempool dumpster to the mempool dumpster
datasource itself. Before, it was on the fetch command. This also
enables a unit test that was ignored before.

---

## ✅ I have completed the following steps:

* [x] Run `make lint`
* [x] Run `make test`
* [x] Added tests (if applicable)

ZanCorDX commented:
This looks like a copy/paste from EnvOrInplaceValue, can you try to refactor it?
Ideally "env:" should be a constant.

ferranbt committed:
feat: Use alloy types for mevboost client (#15)

## 📝 Summary

This PR replaces the custom types in the MevBoost client with the
counterparts in the alloy library.

---

## ✅ I have completed the following steps:

* [x] Run `make lint`
* [x] Run `make test`
* [x] Added tests (if applicable)

ferranbt committed:
feat: Derive builder domain from beacon api (#35)

## 📝 Summary

Builds on top of #34.

This PR removes the flag `cl_network_config_dir` which is used to load
the consensus network spec to derive the builder signing domain.
Instead, it uses the beacon clients to retrieve the required fields.

---

## ✅ I have completed the following steps:

* [x] Run `make lint`
* [x] Run `make test`
* [x] Added tests (if applicable)

MoeMahhouk committed:
chore: Refactor and code improvements

ZanCorDX committed:
Exclude failed or fully dropped orders (#68)

## 📝 Summary

For merged sbundles, we were wrongly reporting as included failed or
fully dropped subbundles.
A fully dropped sub bundle is a bundle with all droppable children
(TxRevertBehavior::AllowedExcluded) that when executed dropped
everything so no tx was included.

## 💡 Motivation and Context

We saw some weird stuff on our DB.

## ✅ I have completed the following steps:

* [x] Run `make lint`
* [x] Run `make test`
* [x] Added tests (if applicable)

---------

Co-authored-by: Ryan Schneider <ryanleeschneider@gmail.com>

ZanCorDX committed:
ignore test until we fix it (#70)

## 📝 Summary

The test has some race condition and fails from time to time.

## 💡 Motivation and Context

I love making PRs and want to break the Guinness record of PR making.

ZanCorDX committed:
Comment corrections (#72)

## 📝 Summary

Missing corrections I missed to add on a previous PR
(https://github.com/flashbots/rbuilder/pull/53).

---------

Co-authored-by: Alejo Salles <alejosalles@gmail.com>

ZanCorDX committed:
bump alloy to 0.2 (#71)

## 📝 Summary

alloy dependencies changed from 0.1 to 0.2

## 💡 Motivation and Context

Newer dependencies help to build around the rbuilder with fewer
problems.
Also 0.2 comes on blue which highlights my eyes.

## ✅ I have completed the following steps:

* [x] Run `make lint`
* [x] Run `make test`
* [ ] Added tests (if applicable)

metachris commented:
@MoeMahhouk could you rebase please?

ryanschneider committed:
fix(jsonrpc): Return tx hash from eth_sendRawTransaction (#74)

## 📝 Summary

Have `rbuilder` correctly return the transaction hash in response to
`eth_sendRawTransaction`.

## 💡 Motivation and Context

<!--- (Optional) Why is this change required? What problem does it
solve? Remove this section if not applicable. -->

I noticed in local testing that `cast send` always returns null.

NOTE: the same is true of the other RPC methods `eth_sendBundle` etc.
but I figured I'd leave those as a Good First Issue for others.

---

## ✅ I have completed the following steps:

* [x] Run `make lint`
* [x] Run `make test`
* [ ] Added tests (if applicable)

dvush committed:
Fix consistency check for merged sbundles (#82)

There were situations where we had merged sbundles that looked like this

bundle1:
   user_tx, allow_revert: no
   backrun1

bundle 2:
  user_tx: allow_revert yes
  backrun

merged sbundle is concat of these two

and consistency check would think that user_tx wat not allowed to revert
enven though there is a situation where we include bundle2

MoeMahhouk committed:
chore: rebase and minor fix for comma separated urls from env variable

MoeMahhouk committed:
Merge branch 'develop' into env-var-cl-node

MoeMahhouk committed:
chore: fix lint issues

metachris commented on crates/rbuilder/src/building/block_orders/share_bundle_merger.rs:
```diff
@@ -214,12 +214,12 @@ impl<SinkType: SimulatedOrderSink> MultiBackrunManager<SinkType> {
-/// We ONLY use this special flow for orders with some particular structure (see [`ShareBundleMerger::break_down_bundle`]).
-/// SBundle not complying with that will pass through and use the standard handling.
-/// @Pending evaluate if we can avoid to send changes no every order update and "flush" changes all together (since orders are usually processed on batches)
+///   Merging example: SBundle(Bundle(tx)+br1) + SBundle(Bundle(tx)+br2) will become SBundle(Bundle(Bundle(tx)+br1),Bundle(Bundle(tx)+br2)).
```
why adding these spaces?

MoeMahhouk replied:
When I rebased and solved the conflicts. The lint pipeline blocked the merging due to those errors.
Here you can see them
https://github.com/flashbots/rbuilder/actions/runs/10141455345/job/28038732480

PS. interesting though that those files were not introduced by my PR but still blocked my PR to merge.
Also the make lint locally didn't show me to fix them. But I did them manually just to make the PR to progress forward
